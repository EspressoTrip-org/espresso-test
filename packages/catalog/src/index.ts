export * from './stores/CardinalCatalogStore';
export * from './presenters/JSONSchemaPresenter';
export * from './entities/CardinalModelEntity';
export * from './entities/CardinalModelLabelEntity';
export * from './entities/CardinalModelActionEntity';
export * from './definitions/Models';
export * from './definitions/models/index';
