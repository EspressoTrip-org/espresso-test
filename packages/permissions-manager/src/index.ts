export * from './permissions-manager';
export * from './cache-managers';
export * from './event-streams';
export * from './token-decoder';
