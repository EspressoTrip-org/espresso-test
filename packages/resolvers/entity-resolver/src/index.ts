export * from './definitions';
export * from './resolver';
export * from './cache';
