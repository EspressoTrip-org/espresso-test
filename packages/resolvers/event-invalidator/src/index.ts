export * from './topic-modules';
export * from './invalidator';
