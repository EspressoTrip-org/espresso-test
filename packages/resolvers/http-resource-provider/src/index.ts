import * as entity_resolver from '@journeyapps-platform/recon-entity-resolver';
import * as micro_tracing from '@journeyapps-platform/micro-tracing';
import * as micro_logger from '@journeyapps-platform/micro-logger';
import * as sdk from '@journeyapps-platform/sdk-common';

const logger = micro_logger.createLogger('http-resource-provider');

export type Provider = {
  models: string[];
  endpoint: string;
};

export type ProviderConfig = {
  providers: Provider[];
};

export type CreateHttpResourceProviderParams = ProviderConfig & {
  client?: sdk.NodeNetworkClient;
  headers?: sdk.HeaderCollector;
};

/**
 * Creates an HTTP resource provider for collecting document resources from arbitrary endpoints
 */
export const createHttpResourceProvider = (
  params: CreateHttpResourceProviderParams
): entity_resolver.ResourceProvider => {
  const client = params.client || sdk.createNodeNetworkClient();

  return {
    models: params.providers.map((provider) => provider.models).flat(),
    async fetch(model, ids) {
      return micro_tracing.trace('fetch-resources', async (span) => {
        const provider = params.providers.find((provider) => {
          return provider.models.includes(model);
        });
        if (!provider) {
          throw new Error(`no provider configured for the model '${model}'`);
        }

        span.setAttribute('endpoint', provider.endpoint);

        const res = await client.request(provider.endpoint, {
          headers: params.headers,
          method: sdk.METHOD.POST,
          body: {
            model: model,
            ids: ids
          },
          retryable: true
        });

        /**
         * We want to throw here to ensure the recon DB never gets out of sync.
         *
         * If we ran into 5xx errors then the `retryable: true` option above should handle those. If we burn
         * through all retries before whatever network error is causing the 5xx then further failures should
         * block the consumer that is calling invalidate.
         *
         * If this throws other errors like 4xx errors this this normally indicates configuration errors and
         * we want this to block service boot. In other words the `.init()` will end up throwing
         */
        if (res.response.status >= 300) {
          logger.error(`Failed to collect resources from provider`, {
            status: res.response.status,
            response: await res.response.text()
          });
          throw new Error(`Received status code ${res.response.status}. Could not collect resource`);
        }

        const docs = await res.decode<entity_resolver.Document[]>();
        if (!docs) {
          return [];
        }

        span.setAttributes({
          total_docs: docs.length
        });

        return docs;
      });
    }
  };
};
