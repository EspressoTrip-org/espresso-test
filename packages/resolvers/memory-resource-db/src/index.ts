export * from './referential-clone';
export * from './transactions';
export * from './definitions';
export * from './database';
export * from './queries';
