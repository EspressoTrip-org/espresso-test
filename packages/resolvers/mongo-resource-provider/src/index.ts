import * as entity_resolver from '@journeyapps-platform/recon-entity-resolver';
import * as mongo from 'mongodb';

type ModelConfig = {
  collection: mongo.Collection<any>;
  projection?: object;
};

export type CreateHttpResourceProviderParams = {
  models: Record<string, ModelConfig>;
};

export const createMongoResourceProvider = (
  params: CreateHttpResourceProviderParams
): entity_resolver.ResourceProvider => {
  return {
    models: Object.keys(params.models),
    async fetch(model, ids) {
      const config = params.models[model];
      if (!config) {
        throw new Error(`cannot provide for model ${model}`);
      }

      const pipeline = [];

      if (ids) {
        pipeline.push({
          $match: {
            _id: {
              $in: ids.map((id) => new mongo.ObjectId(id))
            }
          }
        });
      }

      pipeline.push({
        $project: {
          id: {
            $convert: {
              input: '$_id',
              to: 'string'
            }
          },
          _id: 0,
          ...(config.projection || {})
        }
      });

      return await config.collection.aggregate<entity_resolver.Document>(pipeline).toArray();
    }
  };
};
