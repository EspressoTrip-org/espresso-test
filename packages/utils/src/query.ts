import * as cardinal from '@journeyapps-platform/types-cardinal';
import type * as mongo from 'mongodb';
import * as bson from 'bson';

export type IFilterQueryConstructor<T> = (permission: cardinal.Permission) => mongo.Filter<T> | null | undefined;

/**
 * Utility to augment a MongoDB query with filter criteria generated from a set of permissions. This ensures the permission criteria are
 * not overridden by the provided query and that it fails closed in the case of no permissions.
 *
 * A query constructor must be provided which should return a MongoDB FilterQuery from a single given permission. This query will be $or'd
 * with all other constructed queries and then $and'ed with the given to-be-wrapped query.
 */
export const createFilterQueryFromPermissionSet = <T>(
  permissions: cardinal.Permission[],
  constructor: IFilterQueryConstructor<T>,
  query: mongo.Filter<T>
) => {
  const { allow, deny } = cardinal.splitByEffect(permissions);

  const constructFilterCriteria = (permissions: cardinal.Permission[], operand: '$or' | '$nor') => {
    return permissions.reduce((filter: mongo.Filter<any>, permission) => {
      if (!filter[operand]) {
        filter[operand] = [];
      }
      const query = constructor(permission);
      if (!query) {
        return filter;
      }
      filter[operand]!.push(query as any);
      return filter;
    }, {});
  };

  const deny_criteria = constructFilterCriteria(deny, '$nor');
  const allow_criteria = constructFilterCriteria(allow, '$or');

  // Fail closed if no permissions went into generating this query
  if (!allow_criteria.$or || allow_criteria.$or.length === 0) {
    return null;
  }

  const criteria = {
    $and: [allow_criteria, query]
  };

  if (deny_criteria.$nor?.length) {
    criteria.$and.push(deny_criteria);
  }

  return criteria as mongo.Filter<T>;
};

/**
 * Given a field from a permission return a mongo QuerySelector based on its value.
 *
 * - If it is a standard string, return it as-is
 * - If it is a wildcard ('*') return an `{$exists: true}` filter
 * - If it is empty/undefined/null - return an `{$exists: false}` filter
 */
export const matchWithWildcards = <T>(field?: cardinal.StringOr): mongo.FilterOperators<T> | string[] => {
  if (field == null) {
    return {
      $exists: false
    };
  }
  const fields = cardinal.castToArray(field);
  if (fields.includes('*')) {
    return {
      $exists: true
    };
  }
  return fields;
};

/**
 * The same as matchWithWildcards except that if the result is still a string it is assumed
 * to be an ObjectId and converted as such.
 *
 * If the `loose` parameter is true then an $in query selector is returned, matching both
 * the `string` and `ObjectId` forms of the field (loosely typed)
 */
export const matchIdWithWildcards = (
  field?: cardinal.StringOr,
  loose: boolean = false
): mongo.FilterOperators<any> | bson.ObjectId => {
  const match = matchWithWildcards(field);
  if (Array.isArray(match)) {
    return {
      $in: match
        .map((field) => {
          if (loose) {
            if (!bson.ObjectId.isValid(field)) {
              return [field];
            }
            return [field, new bson.ObjectId(field)];
          } else {
            if (!bson.ObjectId.isValid(field)) {
              return [];
            }
            return [new bson.ObjectId(field)];
          }
        })
        .flat()
    };
  }
  return match;
};

export type OrgScopedResource = {
  org_id?: string | bson.ObjectId;
};

export type LabelQueryConstructor = (
  key: string,
  value: cardinal.StringOr,
  labels: cardinal.SelectorLabels
) => mongo.FilterOperators<string>;
export const defaultLabelQueryConstructor: LabelQueryConstructor = (key: string, value: cardinal.StringOr) => {
  return {
    [`labels.${key}`]: {
      $in: cardinal.castToArray(value)
    }
  };
};

/**
 * A top-level query constructor that assumes the datum schema contains an `org_id` field. This will construct a
 * mongo query from a permission, mapping the permission scope to the `org_id` field and the first resource id
 * to the _id field
 *
 * Does not work for nested permissions. Use the underlying query constructor for nested permissions
 */
export const orgScopedResourceQueryConstructor =
  (
    label_query_constructor: LabelQueryConstructor = defaultLabelQueryConstructor
  ): IFilterQueryConstructor<OrgScopedResource> =>
  (permission) => {
    if (cardinal.isIDResourceSelector(permission.resource.selector)) {
      return {
        _id: matchIdWithWildcards(permission.resource.selector.id),
        org_id: matchIdWithWildcards(permission.resource.scope, true)
      };
    }

    // We map over all label pairs passing them to the label query constructor. This is done
    // as such to allow the caller to provide a custom mapper for labels (labels might not
    // map directly to a `labels.<label>` property on the resource).
    const { labels } = permission.resource.selector;
    const label_queries = Object.keys(labels)
      .map((key) => {
        return label_query_constructor(key, labels[key], labels);
      })
      .filter((query) => query);

    if (label_queries.length === 0) {
      return;
    }

    return {
      $and: [
        {
          org_id: matchIdWithWildcards(permission.resource.scope, true)
        },
        ...label_queries
      ]
    };
  };

export const createOrgScopedFilterQueryFromPermissionSet = <T>(
  permissions: cardinal.Permission[],
  params: {
    models?: string[];
    actions?: string[];
    query: mongo.Filter<T>;
    map_labels?: LabelQueryConstructor;
  }
) => {
  const filters = [params.actions ? cardinal.requireAction(...params.actions) : cardinal.requireAction('read')];
  if (params.models) {
    filters.push(cardinal.requireModel(...params.models));
  }
  const filtered_permissions = cardinal.filterPermissionsBy(permissions, ...filters);

  return createFilterQueryFromPermissionSet<T>(
    filtered_permissions,
    orgScopedResourceQueryConstructor(params.map_labels) as any,
    params.query
  );
};
