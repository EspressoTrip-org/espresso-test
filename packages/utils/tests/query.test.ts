import { describe, test, expect } from 'vitest';

import * as cardinal from '@journeyapps-platform/types-cardinal';
import * as permission_utils from '../src/query';
import * as mongo from 'mongodb';

describe('query construction', () => {
  type Model = {
    id: string;
    org_id: string;
    name: string;
  };

  const query: mongo.Filter<Model> = {
    name: /^app.*/
  };

  const permissions: cardinal.Permission[] = [
    {
      action: 'read',
      resource: {
        scope: '5ee204dcae8864186979d77f',
        selector: {
          model: 'app',
          id: '5ee204dcae8864186979d77f'
        }
      }
    },
    {
      action: 'read',
      resource: {
        scope: '5ee204ddae8864186979d780',
        selector: {
          model: 'app',
          id: '5ee204ddae8864186979d780'
        }
      }
    }
  ];

  test('construct mongodb filter query from permissions', () => {
    const filter = permission_utils.createFilterQueryFromPermissionSet(
      permissions,
      (permission) => {
        if (permission.resource.selector.model !== 'app') {
          return;
        }
        if (!cardinal.isIDResourceSelector(permission.resource.selector)) {
          return;
        }
        return {
          org_id: {
            $in: cardinal.castToArray(permission.resource.scope!)
          },
          id: {
            $in: cardinal.castToArray(permission.resource.selector.id)
          }
        };
      },
      query
    );

    expect(filter).toMatchSnapshot();
  });

  test('construct mongodb filter query from permissions using org scoped tools', () => {
    const filter = permission_utils.createOrgScopedFilterQueryFromPermissionSet(permissions, {
      models: ['app'],
      query
    });

    expect(filter).toMatchSnapshot();
  });

  test('construct query with wildcards', () => {
    const permissions: cardinal.Permission[] = [
      {
        action: 'read',
        resource: {
          scope: '5ee204dcae8864186979d77f',
          selector: {
            model: 'app',
            id: '*'
          }
        }
      }
    ];

    const filter = permission_utils.createOrgScopedFilterQueryFromPermissionSet(permissions, {
      models: ['app'],
      query
    });

    expect(filter).toMatchSnapshot();
  });

  test('construct query with action and model wildcards', () => {
    const permissions: cardinal.Permission[] = [
      {
        action: '*',
        resource: {
          scope: '5ee204dcae8864186979d77f',
          selector: {
            model: '*',
            id: '*'
          }
        }
      }
    ];

    const filter = permission_utils.createOrgScopedFilterQueryFromPermissionSet(permissions, {
      models: ['app'],
      actions: ['read'],
      query
    });

    expect(filter).toMatchSnapshot();
  });

  test('construct query with scopeless selector', () => {
    const permissions: cardinal.Permission[] = [
      {
        action: 'read',
        resource: {
          selector: {
            model: 'app',
            id: '5ee204dcae8864186979d77f'
          }
        }
      }
    ];

    const filter = permission_utils.createOrgScopedFilterQueryFromPermissionSet(permissions, {
      models: ['app'],
      query
    });

    expect(filter).toMatchSnapshot();
  });

  test('construct query with scoped and scopeless selector', () => {
    const permissions: cardinal.Permission[] = [
      {
        action: 'read',
        resource: {
          selector: {
            model: 'app',
            id: '5ee204dcae8864186979d77f'
          }
        }
      },
      {
        action: 'read',
        resource: {
          scope: '5ee204dcae8864186979d77f',
          selector: {
            model: 'app',
            id: '5ee204ddae8864186979d780'
          }
        }
      }
    ];

    const filter = permission_utils.createOrgScopedFilterQueryFromPermissionSet(permissions, {
      models: ['app'],
      query
    });

    expect(filter).toMatchSnapshot();
  });

  test('construct query from relevant permissions', () => {
    const permissions: cardinal.Permission[] = [
      {
        action: 'read',
        resource: {
          selector: {
            model: 'app',
            id: '5ee204dcae8864186979d77f'
          }
        }
      },
      {
        action: 'write',
        resource: {
          scope: '5ee204dcae8864186979d77f',
          selector: {
            model: 'app',
            id: '5ee204dcae8864186979d77f'
          }
        }
      },
      {
        action: 'read',
        resource: {
          scope: '5ee204dcae8864186979d77f',
          selector: {
            model: 'book',
            id: '5ee204dcae8864186979d77f'
          }
        }
      }
    ];

    const filter = permission_utils.createOrgScopedFilterQueryFromPermissionSet(permissions, {
      models: ['app'],
      query
    });

    expect(filter).toMatchSnapshot();
  });

  test('construct query empty permissions set', () => {
    const permissions: cardinal.Permission[] = [];

    const filter = permission_utils.createOrgScopedFilterQueryFromPermissionSet(permissions, {
      models: ['app'],
      query
    });

    expect(filter).toEqual(null);
  });

  test('should handle malformed ids correctly', () => {
    const permissions: cardinal.Permission[] = [
      {
        action: 'read',
        resource: {
          scope: '123',
          selector: {
            model: 'app',
            id: '456'
          }
        }
      }
    ];

    const filter = permission_utils.createOrgScopedFilterQueryFromPermissionSet(permissions, {
      models: ['app'],
      query
    });

    expect(filter).toMatchSnapshot();
  });
});
