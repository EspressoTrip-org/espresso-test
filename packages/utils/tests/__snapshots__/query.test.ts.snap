// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`query construction > construct mongodb filter query from permissions 1`] = `
{
  "$and": [
    {
      "$or": [
        {
          "id": {
            "$in": [
              "5ee204dcae8864186979d77f",
            ],
          },
          "org_id": {
            "$in": [
              "5ee204dcae8864186979d77f",
            ],
          },
        },
        {
          "id": {
            "$in": [
              "5ee204ddae8864186979d780",
            ],
          },
          "org_id": {
            "$in": [
              "5ee204ddae8864186979d780",
            ],
          },
        },
      ],
    },
    {
      "name": /\\^app\\.\\*/,
    },
  ],
}
`;

exports[`query construction > construct mongodb filter query from permissions using org scoped tools 1`] = `
{
  "$and": [
    {
      "$or": [
        {
          "_id": {
            "$in": [
              "5ee204dcae8864186979d77f",
            ],
          },
          "org_id": {
            "$in": [
              "5ee204dcae8864186979d77f",
              "5ee204dcae8864186979d77f",
            ],
          },
        },
        {
          "_id": {
            "$in": [
              "5ee204ddae8864186979d780",
            ],
          },
          "org_id": {
            "$in": [
              "5ee204ddae8864186979d780",
              "5ee204ddae8864186979d780",
            ],
          },
        },
      ],
    },
    {
      "name": /\\^app\\.\\*/,
    },
  ],
}
`;

exports[`query construction > construct query from relevant permissions 1`] = `
{
  "$and": [
    {
      "$or": [
        {
          "_id": {
            "$in": [
              "5ee204dcae8864186979d77f",
            ],
          },
          "org_id": {
            "$exists": false,
          },
        },
      ],
    },
    {
      "name": /\\^app\\.\\*/,
    },
  ],
}
`;

exports[`query construction > construct query with action and model wildcards 1`] = `
{
  "$and": [
    {
      "$or": [
        {
          "_id": {
            "$exists": true,
          },
          "org_id": {
            "$in": [
              "5ee204dcae8864186979d77f",
              "5ee204dcae8864186979d77f",
            ],
          },
        },
      ],
    },
    {
      "name": /\\^app\\.\\*/,
    },
  ],
}
`;

exports[`query construction > construct query with scoped and scopeless selector 1`] = `
{
  "$and": [
    {
      "$or": [
        {
          "_id": {
            "$in": [
              "5ee204dcae8864186979d77f",
            ],
          },
          "org_id": {
            "$exists": false,
          },
        },
        {
          "_id": {
            "$in": [
              "5ee204ddae8864186979d780",
            ],
          },
          "org_id": {
            "$in": [
              "5ee204dcae8864186979d77f",
              "5ee204dcae8864186979d77f",
            ],
          },
        },
      ],
    },
    {
      "name": /\\^app\\.\\*/,
    },
  ],
}
`;

exports[`query construction > construct query with scopeless selector 1`] = `
{
  "$and": [
    {
      "$or": [
        {
          "_id": {
            "$in": [
              "5ee204dcae8864186979d77f",
            ],
          },
          "org_id": {
            "$exists": false,
          },
        },
      ],
    },
    {
      "name": /\\^app\\.\\*/,
    },
  ],
}
`;

exports[`query construction > construct query with wildcards 1`] = `
{
  "$and": [
    {
      "$or": [
        {
          "_id": {
            "$exists": true,
          },
          "org_id": {
            "$in": [
              "5ee204dcae8864186979d77f",
              "5ee204dcae8864186979d77f",
            ],
          },
        },
      ],
    },
    {
      "name": /\\^app\\.\\*/,
    },
  ],
}
`;

exports[`query construction > should handle malformed ids correctly 1`] = `
{
  "$and": [
    {
      "$or": [
        {
          "_id": {
            "$in": [],
          },
          "org_id": {
            "$in": [
              "123",
            ],
          },
        },
      ],
    },
    {
      "name": /\\^app\\.\\*/,
    },
  ],
}
`;
